### ISSUE TITLE
Question: When is MapControl coming to WinUI 3.x?

### ISSUE THREAD

**Issue by robloo** [microsoft/microsoft-ui-xaml#3024]
I have an app that depends on MapControl which isn't currently supported by WinUI 3.0. When might I expect that 
to be added? 

WinUI 3.0 Preview 2 lists MapControl as a missing platform feature with no timeline given.
https://docs.microsoft.com/en-us/windows/apps/winui/winui3/#missing-platform-features

I don't even need all the advanced functionality -- just something that shows a 2D map and allows pins to be added.

**Comment by robloo** [comment]
Adding @YuliKl who seems to be assigned MapControl-related issues.

**Comment by robloo** [comment]
@YuliKl Any news on this?

**Comment by robloo** [comment]
@YuliKl @StephenLPeters Any updates on this? This is going to end up being a show stopper for me using winUI 3. It's pretty important this gets on a roadmap so plans can be made one way or another.

**Comment by StephenLPeters** [comment]
@Austin-Lamb or @MikeHillberg ?

**Comment by robloo** [comment]
On the latest roadmap this is considered a stretch goal even after 3.0 release. I'm guessing that means this control might not get updated for years -- if ever. What is the alternative? I could use Bing maps hosted in WebView2 if only it was possible to put pins on the map. Bing doesn't support that though. Can we get that team involved and come up with something at least in the interim?

**Comment by p2pbsh** [comment]
Bump on this one. The MapControl is pretty much the centrepiece of my UWP app, seeing it listed as stretch goal of 3.x is slightly ominous in regards of me being able to adopt WinUI. Could someone please elaborate with some more concrete info on the plans for MapControl please?

**Comment by ryandemopoulos** [comment]
Hey @robloo, we don't have a specific timeframe for a Map control, but I suspect it will be a while.   We tried to indicate this in our [public feature roadmap](https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/roadmap.md) (see the last row in the table at the bottom of the page).  I could see us adding it in a 3.x point release after 3.0 ships, but I doubt it would be 3.1 or 3.2, which very likely will be focused on other things like Acrylic, XAML Islands, more WPF parity functionality, tooling innovation, and broadening the set of Windows devices we run on.

While that's definitely no the answer you're looking for, WinUI 2 will continue to be maintained and invested-in so using a system XAML/WinUI 2-powered stack would provide that for you.  Or, since WinUI 3 will offer a lot of hosting flexibility given it's coming interoperability with HWNDs, perhaps there could be other 3rd party options to power a map experience.  Not sure if those alternatives could work for you until we get to a MapControl.

**Comment by robloo** [comment]
@ryandemopoulos Thanks for your response. I do realize it was added to the roadmap some months ago and the latest comments reflect that. You are correct that that's not the answer I was hoping for but it was basically what I was expecting at this point. **You have got to understand how difficult of a place you are putting UWP developers in. I'm now locked out of .net 5, c#9 and WinUI3 for 1-2 years without major rewrites to the app. UWP developers, who invested in Microsofts latest technology, are getting burned in several ways here.**

I guess you have made your WinUI decisions knowing full well it would have some negative impact to those heavily invested in UWP. You made that decision based on UWP usage numbers most likely. Still, I think something should have been done to keep UWP developers from being impacted in this way. They are really the most loyal developers - or those most heavily invested in the tech becoming WinUI3 at this point. 

Reguardless, I have no choice but to pursue rewrites for the MapControl and other missing functionality in WinUI3. I appreciate at least being able to firmly plan on that now.

Its also great you are focusing on closing the feature gap with WPF. Wish it was done 2-3 years ago but better late than never.

Keep up the good work with WinUI3. Its unfortunate I'm in the position I am but overall it will be beneficial to Windows. Hoping a lot of the decisions you are making are also to update the shell and Microsoft's legacy apps.

**Comment by mdtauk** [comment]
You do have to wonder if it would not be better, easier, and more futureproof, to perhaps transition away from a MapControl, and instead offer a WebView/WebView2 implementation, so future updates can come online.  And who knows, maybe even enable Apple Maps and Google Maps support at some point by having a **ServiceProvider** API

**Comment by p2pbsh** [comment]
@ryandemopoulos Thanks for the update. Much like @robloo it's a frustrating time to be a UWP developer, seeing the platform placed on the back burner having committed to it. What I would personally like to see if feasible would be an update of the roadmap that offers clear milestones with migration paths for UWP developers. For example:

- 3.abc - WinUI xaml feature parity with current UWP platform, UWP desktop apps can now be ported to .NET desktop

I'm not particularly bothered by MS deciding to pull the plug on UWP which appears to be the case, makes sense in a lot of ways, just want to be confident the pathway exists to move my work to a WinUI based framework and have an approximate idea of timescales so I can plan ahead accordingly. I'd argue that a complete 'escape route' for UWP devs should really be in place by the time .NET6 drops.

**Comment by leoniDEV** [comment]
@mdtauk  But for example with MapControl you can leverage offline maps!

**Comment by robloo** [comment]
@mdtauk I looked into using a map within webview already. The issue is the APIs dont support really anything. MapControl has a lot of advanced functionality to draw on the map. I only need push pin support in my case; however, I can't even place multiple pins with any online maps provider and control the zoom level. Best I can do with tricks is to place a single pin. No offline support is unfortunate as well but would be tolerable if anything else worked.

MapControls are expensive to implement primarily due to the data. Even using OpenStreetMaps you have to pay for rendered tile server data. Otherwise, adding this functionality, and storage capacity to your company's existing servers is not cheap.

**Comment by robloo** [comment]
It anyone knows of some fancy math or techniques that could be used to calculate the geographical position coordinates of the pixels in a Google or Bing maps there are some ideas. It might be possible to write a new control using a layered technique and manually drawing on top of a webview2 containing the map on the bottom layer. Keeping things positional synronized during movement would also be difficult though. Recalculation when the underlying map changed projection modes would be nearly impossible I suspect.

**Comment by mdtauk** [comment]
> It anyone knows of some fancy math or techniques that could be used to calculate the geographical position coordinates of the pixels in a Google or Bing maps there are some ideas. It might be possible to write a new control using a layered technique and manually drawing on top of a webview2 containing the map on the bottom layer. Keeping things positional synronized during movement would also be difficult though. Recalculation when the underlying map changed projection modes would be nearly impossible I suspect.

If Microsoft were to go the wrapped WebView approach with a future Map Control - it would need to have Offline data support, as well as supporting the drawing of polygons on-top of the map layers.  It would also need to be hardware accelerated.

Chromium can do all this, using PWA tech AFAIK.

**Comment by robloo** [comment]
The link below seems to be the best control available for replacement. Still not sure if the map provider is free though:

https://github.com/Mapsui/Mapsui

**Comment by Mogikan** [comment]
I haven't thought WinUI 3.0 could be incompatible with existing UWP app. Please provide controls which were available in UWP. It is a bad tone to decide to throw away controls! This is a really a stopper problem for my app.

**Comment by wjax** [comment]
I have tumbled here because I wanted to evaluate porting my WPF app to WinUI. I wanted to use the great Map Control from UWP... Seems there is no chance and I will keep using my old but bullet proof WPF Win32 apps.

**Comment by venklar** [comment]
I second all of the above: Map Control is required ASAP!
I work on GIS and for me the lack of a Map control is a blocker to addapting WinUI 3.

Not sure but I think using Uno Platform it's possible to use both WinUI and UWP Map Control?...

**Comment by Mogikan** [comment]
Hi, I see that MapControl disappeared from Roadmap. What does it mean? Are they supported in WinUI 3.0 or they will never be implemented?
https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md

**Comment by DaveAurionix** [comment]
Placing the emphasis a little differently here - the existing UWP MapControl is great.  Native mapping performance on Windows devices, with multiple input devices supported, powerful customisations, lines, shapes, pins and completely customisable styling.  Offline maps support.  It fills a niche.  Yes, it has a couple of bugs here and there (like disappearing labels).  I'd absolutely love to see this ported to WinUI3 (to unblock an app I work on from moving to WinUI3), and maybe have a couple of the bugs fixed too.  I'd really prefer not to go this far, but if anyone at Microsoft does want to employ a software engineer to help get the job done from the UWP source code, DM me!  As an end-user of the UWP MapControl, I understand the end goal of the WinUI3 port too.  I'd love to see a WinUI3 port done by Microsoft, but failing that I'd love to _help_ get it done myself!

**Comment by JORGEGO** [comment]
It is unheard of how the community of developers and companies that opted for UWP for Xamarin Forms and have applications developed with use of map control can be left without a solution and now there is no clear solution, nor plan to offer the MAUI team a map control to integrate it into MAUI, it is incredible and at the same time worrying, the question is: what do we do?!!!!!!!

**Comment by robloo** [comment]
It isn't unheard of. Microsoft has done the same type of thing (clear loss of features and functionality) going from WPF to UWP. In this case there is even less of a concern because there were far fewer UWP apps.

https://github.com/microsoft/microsoft-ui-xaml/issues/3639

**Comment by p2pbsh** [comment]
In the end I swapped out for MAPSUI. Worked well for me, may work for you:

https://github.com/Mapsui/Mapsui

**Comment by JORGEGO** [comment]
> In the end I swapped out for MAPSUI. Worked well for me, may work for you:
> 
> https://github.com/Mapsui/Mapsui

I will test, who is supporting this control?

**Comment by p2pbsh** [comment]
> I will test, who is supporting this control?

It's a community project, but actively maintained, the main guys working on it are pretty on it.

To be fair, the old MS UWP control has some terrible performance issues and would cause random crashes to desktop, which were never fixed. I've found mapsui to be very solid, also far more powerful it terms of functionality once you get your head round how it all bolts together.

**Comment by JORGEGO** [comment]
I am considering using this component, I find it very interesting, I have been testing it and it has a very good performance. I have tried it on Windows, I have to try it on Android and iOS, the question is about Android and iOS does it use the capabilities of Android and iOS, read for these two platforms does it act as Map MAUI control?
The next question is, now it is in preview and there is no cost, is there any kind of licensing planned in the future?

**Comment by robloo** [comment]
On my end MapsUI was used as well (as indicated above). That's possible now that it is MIT licenced. It works with Avalonia now as well which is being moved to away from WinUI.

**Comment by p2pbsh** [comment]
> I am considering using this component, I find it very interesting, I have been testing it and it has a very good performance. I have tried it on Windows, I have to try it on Android and iOS, the question is about Android and iOS does it use the capabilities of Android and iOS, read for these two platforms does it act as Map MAUI control?
> The next question is, now it is in preview and there is no cost, is there any kind of licensing planned in the future?

As robloo says, current beta moves everything to MIT, so project trajectory is away from licensing requirements rather than towards them. 

Not sure exactly what you mean by device capabilities, but I just finished porting an app to MAUI using device geolocation and compass to indicate users position and direction on map and all working nicely on Android and iOS.

**Comment by mfeingol** [comment]
So is it fair to say at this point that WinUI will not get a native MapControl for the foreseeable future?

**Comment by p2pbsh** [comment]
They discussed it in the last WinUI community call, sounded like they still want to do it, but I wouldn't expect seeing it anytime soon.

**Comment by TFreudi1** [comment]
The missing map control in WinUI3 blocks me from porting Xamarin App's to NET. MAUI !

**Comment by Panda-Sharp** [comment]
> The missing map control in WinUI3 blocks me from porting Xamarin App's to NET. MAUI !

this...and Ink Canvas Control

**Comment by robloo** [comment]
This is still valid for a lot of people. Windows Community toolkit had to end up wrapping a web control. And we also have Mapsui. But the WinUI team needs to make a decision and report finally if this control will ever come.

**Comment by mfeingol** [comment]
> This is still valid for a lot of people. Windows Community toolkit had to end up wrapping a web control. And we also have Mapsui. But the WinUI team needs to make a decision and report finally if this control will ever come.

Agreed. MapsUI is missing functionality like 3D rendering that are key use cases in our UWP apps. It's frustrating to not have any signs of life on this topic.

**Comment by Panda-Sharp** [comment]
> > This is still valid for a lot of people. Windows Community toolkit had to end up wrapping a web control. And we also have Mapsui. But the WinUI team needs to make a decision and report finally if this control will ever come.
> 
> Agreed. MapsUI is missing functionality like 3D rendering that make up key use cases in our UWP apps. It's frustrating to not have any signs of life on this topic.

It's frustrating the sign of life of winui in general 😓

**Comment by Mogikan** [comment]
I should consider Flutter for Windows development next time, at least they have maps on Windows. I feel Silverlighted every time when it comes to Microsoft technologies. WinRT (and winui) was a mistake and it should die with windows phone. WPF is much better for development.

**Comment by Mogikan** [comment]
Maps will appear when pigs fly. Maybe they are waiting to close winui project since noone uses it. Or maybe it is a low priority project for entuasiasts. I believe that it was possible to implement maps and MS developers feel ashaimed that they can do nothing because of the poor management they have. Do you know the story that in xamarin team managers were so lazy that hopeless developer implemented catalyst feature himself (he waited two years in similar ping pong chat) spending one or two weeks to make it work (prototype) and only then management added this feature to roadmap. The reason was probably to wait extra 10 years until all intel macs will become obsolete themselves. Maybe they feel the same about winui. But you will never know until being brutally "silverlighted again". What is the point of this platform if WPF project was unable to die after 15 years of being abandoned. Windows mobile is a limited platform with too many restrictions and windows store is a flea market no one knows about.

**Comment by bricelam** [comment]
https://github.com/microsoft/microsoft-ui-xaml/issues/6293#issuecomment-1659818942 gave me a bit of hope. I don't pretend to know this control would actually be implemented, but hosting the UWP map control inside a XAML island was how they had it working in WinForms and WPF for .NET Core 3. Sadly, .NET Core 3 is out of support and XAML islands stopped working in the next version of .NET.

**Comment by rmarinho** [comment]
So on MAUI we added to the toolkit a WinUI Map that is a wrapper on a WebPage with bing maps. So if anyone wants to grab some inspiration how to map from c# to javascript and back .. for some basic stuff. 

https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui.Maps/Handler/Map/MapHandler.Windows.cs

**Comment by duncanmacmichael** [comment]
Closing this question as the Maps Control is on track to be delivered in the WinAppSDK 1.5, per the currently posted roadmap:

https://github.com/microsoft/WindowsAppSDK/blob/main/docs/roadmap.md

For any questions about the new maps control or suggestions when it arrives, please feel free to open a new issue. Thanks!

**Comment by robloo** [comment]
I'm surprised this is coming, even if much later than it was needed. Great work on this!

**Comment by hez2010** [comment]
> Closing this question as the Maps Control is on track to be delivered in the WinAppSDK 1.5, per the currently posted roadmap:
> 
> [microsoft/WindowsAppSDK@`main`/docs/roadmap.md](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/roadmap.md?rgh-link-date=2023-12-09T02%3A15%3A22Z)
> 
> For any questions about the new maps control or suggestions when it arrives, please feel free to open a new issue. Thanks!

Please don't ship the WebView based MapControl, what we expect is a native MapControl just like the one in UWP. And we expect the MapControl to work even if there's no Internet access so that we can still use the offline map data which has been downloaded in advance.

**Comment by moshegutman** [comment]
Please don't ship a non-native control. The Azure Maps base map doesn't come close to the Bing Maps base map. Also, for those of us using the UWP control under a Bing Maps Enterprise Agreement, I don't see how this new control could possibly fulfill the contractual obligations of that agreement.

**Comment by mfeingol** [comment]
@hez2010: two question:

1. Why do we think it won't be a native MapControl?
2. How exactly would one use the UWP MapControl in a WinUI app? Last I checked that project doesn't work beyond .NET 5...

**Comment by tipa** [comment]
> 1. Why do we think it won't be a native MapControl?

https://github.com/microsoft/microsoft-ui-xaml/blob/winui3/release/1.5-stable/controls/dev/MapControl/MapControl.xaml

I agree that a WebView map control just won't do it. I haven't tested it myself but I have a very, very strong feeling that it will be lacking a lot of features compared to the UWP MapControl (e.g. binding a collection that renders XAML on the map) and also performance will be subpar

**Comment by mfeingol** [comment]
Thanks, @tipa. Looking at the IDL, the API looks extremely basic:

https://github.com/microsoft/microsoft-ui-xaml/blob/winui3/release/1.5-stable/controls/dev/MapControl/MapControl.idl

A deal-breaker for me would be the inability to render MapTileSources. This doesn't bode well...

**Comment by mfeingol** [comment]
Well, I poked around 1.5.240205001-preview1 for a few minutes. It's worse than I expected.

I was unable to get MapControl working beyond a blue Azure screen, despite trying several different Azure/Bing Maps API keys. So I can't comment on how things look or what sorts of controls are in the HTML/JS.

API surface area-wise, it's a complete disaster. I'm not even sure how the Maui folks will implement their relatively limited surface area.

**MapTileSource** is gone, so no custom tiles are possible.

**MapStyle** is gone. No way to programmatically select map type.

There's a MapLayer property that does something, but it's unclear how to use it. There's a MapElementsLayer subtype which appears to hold MapElements instead of the usual MapElements property.

The **Children** property is also gone, so no XAML elements can be used. Unsurprisingly.

**MapRouteView** is gone.

**MapPolyline** and **MapPolygon** are gone. **MapIcon** is still alive but has no Title, Image, ZIndex, NormalizedAnchorPoint or anything hinting at customization.

CenterChanged and ZoomLevelChanged are gone. MapElementPointerEntered and MapElementPointerExited are gone. FindMapElementsAtOffset, GetVisibleRegion and TrySetViewBoundsAsync are all gone. Good luck actually using this MapControl for anything sophisticated.

ColorScheme and MapProjection are gone. ZoomInteractionMode, TiltInteractionMode, PanInteractionMode and RotateInteractionMode are gone.

And this is just the surface area my app uses. At current course and speed, WinUI is no longer a viable platform for me and I will need to look at third party offerings.

**Comment by tipa** [comment]
@mfeingol Thanks for testing & reporting.
What a joke... I won't even care creating feature requests for these missing (and essential) features, Microsoft obviously has no interest in providing a usable mapping control

**Comment by mfeingol** [comment]
I'd be curious to see if someone can figure out how to get it working at least... i.e. which MapServiceToken it will accept.

Beyond that... it's Preview 1, the core JS control clearly has more capabilities, those might light up in future previews... But I'm already sensing that this won't be able to fully replace UWP MapControl for my uses.

MapsUI seems like the most advanced third party control, staying within WinUI. Unfortunately it's missing the nice 3D views that more advanced controls have. That will be a big loss for my scenarios.

**Comment by dotMorten** [comment]
@mfeingol just to toot my own horn: [`Esri.ArcGISRuntime.WinUI`](https://www.nuget.org/packages/Esri.ArcGISRuntime.WinUI) fully supports both 2D and 3D maps and is using native directx rendering into WinUI's swapchain. I realize it might be overkill for your needs, but it'll do what the UWP control you used did and then some.

**Comment by mfeingol** [comment]
@dotMorten: I hadn't realized there was a free beer license for Esri's map control, so that merits further exploration. It looks like there's lot of functionality in there!

One thing I don't love, a priori, is the distinction between Maps and Scenes. A really nice feature of the UWP control is the seamless transition between 2D and 3D views fed by elevation data. Can this control do something similar?

**Comment by dotMorten** [comment]
@mfeingol let's take this outside here. Ping me on twitter, discord, etc on this username and I'll be happy to help out (this goes for everyone)

**Comment by dotMorten** [comment]
@mfeingol btw the blue screen issue is reported here: https://github.com/microsoft/microsoft-ui-xaml/issues/9324 This issue seems to still exist in final v1.5 too
